### PR TITLE
Functional test module

### PR DESCRIPTION
**Pull request by xepozz**
Ready for merge

### CODE REVIEW

File: tests/functional/RouteTest.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public function testContainer()
+    {
+        self::$application = null;
+        $container = $this->getContainer();
+
+        $this->assertInstanceOf(ContainerInterface::class, $container);
+        $this->assertTrue($container->has(LoggerInterface::class));
+    }
+
+    private function doRequest(string $method, string $url): ?ResponseInterface
```
Comment (samdark) [inline tests/functional/RouteTest.php:46]: Do you mean moving these to a trait or to a base test case as a protected method?

File: tests/functional/RouteTest.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        $method = 'GET';
+        $url = '/';
+
+        $response = $this->doRequest($method, $url);
+
+        $this->assertInstanceOf(ResponseInterface::class, $response);
+
+        $stream = $response->getBody();
+        $stream->rewind();
+        $content = $stream->getContents();
```
Comment (samdark) [inline tests/functional/RouteTest.php:29]: That's a bit too verbose. Maybe worth adding a shortcut.

**Review by samdark — APPROVED** [review]
I like the concept 👍

**Comment by xepozz** [comment]
Here two options how to use this functionality:
1. [Using composition](https://github.com/yiisoft/demo-api/blob/df6243b4974aff23e5f7adf041cc43c54b50fb23/tests/functional/UserControllerTest.php#L25)
2. [Using inheritance](https://github.com/yiisoft/demo-api/blob/df6243b4974aff23e5f7adf041cc43c54b50fb23/tests/functional/RouteTest.php#L18)

The second one available thanks to decomposition [test functionality](https://github.com/yiisoft/demo-api/blob/df6243b4974aff23e5f7adf041cc43c54b50fb23/tests/functional/FunctionalTester.php) from [test case](https://github.com/yiisoft/demo-api/blob/df6243b4974aff23e5f7adf041cc43c54b50fb23/tests/functional/FunctionalTestCase.php).

**Comment by xepozz** [comment]
https://github.com/yiisoft/yii-testing/pull/2
Here is a fix for the use of `yii-dev-tool`

File: tests/functional/RouteTest.php (hunk 2 above)
Comment (xepozz) [reply tests/functional/RouteTest.php:26]: https://github.com/yiisoft/yii-testing/issues/4

File: src/InfoController.php [hunk 3]
```diff
@@ -47,6 +51,6 @@ class InfoController
      */
     public function index(DataResponseFactoryInterface $responseFactory): ResponseInterface
     {
-        return $responseFactory->createResponse(['version' => '3.0', 'author' => 'yiisoft']);
+        return $responseFactory->createResponse(['version' => $this->versionProvider->version, 'author' => 'yiisoft']);
```
Comment (arogachev) [inline src/InfoController.php:54]: Did not fully understand the necessity of it. Also why not extract author as well then?

File: src/VersionProvider.php [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App;
+
+class VersionProvider
+{
+    public function __construct(public $version = '3.0')
```
Comment (arogachev) [inline src/VersionProvider.php:9]: Maybe add a method for possibility of getting it dynamically? Or it's redundant for now?

**Review by arogachev — COMMENTED** [review]
:+1:

File: src/InfoController.php (hunk 3 above)
Comment (xepozz) [reply src/InfoController.php:54]: That's just showcase how to mock dependencies

**Pull request merged**